receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,539 @@
+func (c *snmpClient) Close() error {
+	return c.client.Close()
+}
+
+// GetScalarData retrieves scalar metrics from passed in scalar OIDs. The returned data
+// is then also passed into the provided function.
+// Note: These OIDs must all end in ".0" for the SNMP GET to work correctly
+func (c *snmpClient) GetScalarData(oids []string, processFn processFunc) error {
+	// Nothing to do if there are no OIDs
+	if len(oids) == 0 {
+		return nil
+	}
+
+	// Group OIDs into chunks based on the max amount allowed in a single SNMP GET
+	chunkedOIDs := chunkArray(oids, c.client.GetMaxOids())
+	getOIDsSuccess := false
+	processOIDSuccess := false
+
+	// For each group of OIDs
+	for _, oids := range chunkedOIDs {
+		// Note: Not implementing GetBulk as I don't think it would work correctly for the current design
+		packets, err := c.client.Get(oids)
+		if err != nil {
+			c.logger.Warn("Problem with GET OIDs", zap.Error(err))
+			// Prevent getting stuck in a failure where we can't recover
+			if strings.Contains(err.Error(), "request timeout (after ") {
+				c.Close()
+				c.Connect()
+			}
+			continue
+		}
+		// Keep track of if at least one GET successfully returned data
+		getOIDsSuccess = true
+
+		// For each piece of data in a returned packet
+		for _, data := range packets.Variables {
+			// If there is no value, then ignore
+			if data.Value == nil {
+				c.logger.Warn(fmt.Sprintf("Data for OID: %s not found", data.Name))
+				continue
+			}
+			// Convert data into the more simplified data type
+			snmpData := c.convertSnmpPDUToSnmpData(data)
+			// If the value type is not supported, then ignore
+			if snmpData.valueType == NotSupported {
+				c.logger.Warn(fmt.Sprintf("Data for OID: %s not a supported type", data.Name))
+				continue
+			}
+			// Process the data
+			if err := processFn(snmpData); err != nil {
+				c.logger.Warn(fmt.Sprintf("Problem with processing data for OID: %s", snmpData.oid), zap.Error(err))
+				continue
+			}
+			// Keep track of if at least one set of GET data was successfully processed
+			processOIDSuccess = true
+		}
+	}
+
+	// Return specialized error messages if we failed to retrieve or process any data here.
+	if !getOIDsSuccess {
+		return errNoGetOIDs
+	}
+
+	if !processOIDSuccess {
+		return errNoProcessGetOIDs
+	}
+
+	return nil
+}
+
+// GetIndexedData retrieves indexed metrics from passed in column OIDs. The returned data
+// is then also passed into the provided function.
+func (c *snmpClient) GetIndexedData(oids []string, processFn processFunc) error {
+	// Nothing to do if there are no OIDs
+	if len(oids) == 0 {
+		return nil
+	}
+
+	walkOIDsSuccess := false
+
+	// For each column based OID
+	for _, oid := range oids {
+		// Create a walkFunc which is a required argument for the gosnmp Walk functions
+		walkFn := func(data gosnmp.SnmpPDU) error {
+			// If there is no value, then stop processing
+			if data.Value == nil {
+				return fmt.Errorf("Data for OID: %s not found", data.Name)
```

Comment:
nit: Go errors should not start with capital letters

Reply:
@cpheps oops, yeah i missed this one. i'll fix

Reply:
fixed